tests/pipeline_test.py

```diff
@@ -1208,3 +1208,124 @@ def process(
 def test_html_repr() -> None:
     pipeline = sl.Pipeline([make_int], params={float: 5.0})
     assert isinstance(pipeline._repr_html_(), str)
+
+
+def test_does_not_allow_type_argument_outside_of_constraints_flat() -> None:
+    T = TypeVar('T', int, float)
+
+    @dataclass
+    class M(Generic[T]):
+        value: T
+
+    def p1(value: T) -> M[T]:
+        return M(value)
+
+    pipeline = sl.Pipeline((p1,))
+    pipeline[str] = 'abc'
+    pipeline[int] = 123
+
+    pipeline.get(M[int])
+
+    with pytest.raises(sl.handler.UnsatisfiedRequirement):
+        pipeline.get(M[str])  # type: ignore[type-var]
```

Comment:
Can we define `M` differently (using a different `T`) such that `mypy` passes on this?

Reply:
Yeah we could, but I think it would make the test more confusing. The point of the test is that we're trying to request a type that is outside of the constraints of `T`, I think the `#ignore` helps to communicate that.

Reply:
On the contrary! Remember how *I* was confused, because `M[str]` is illegal. But you then went on to explain to be that what matters is what the provider can provide. I think you must make `M[str]` legal, but keep the provider as it is, i.e., it cannot provider `M[str]`.